### PR TITLE
Remote Address rate limiting

### PR DESCRIPTION
**Pull request by ccaraman**

### CODE REVIEW

**Comment by ccaraman** [comment]
@lyft/network-team

File: docs/configuration/http_filters/rate_limit_filter.rst [hunk 1]
```diff
@@ -87,6 +87,20 @@ descriptor is sent as well:
...
+
+.. code-block:: json
+
+  {
+    "type": "remote_address"
+  }
+
+The following descriptor is sent using the trusted address from `x-forwarded-for <config_http_conn_man_headers_x-forwarded-for>`:
+
+    * ("remote_address", "<:ref:`trusted address from x-forwarded-for <config_http_conn_man_headers_x-forwarded-for>`>")
```
Comment (mattklein123) [inline docs/configuration/http_filters/rate_limit_filter.rst]: send key form as well as this one like the header one.

File: docs/configuration/http_filters/rate_limit_filter.rst [hunk 2]
```diff
@@ -87,6 +87,20 @@ descriptor is sent as well:
...
+Remote Address
+^^^^^^^^^^^^^^
+
+.. code-block:: json
+
+  {
+    "type": "remote_address"
+  }
+
+The following descriptor is sent using the trusted address from `x-forwarded-for <config_http_conn_man_headers_x-forwarded-for>`:
```
Comment (ccaraman) [inline docs/configuration/http_filters/rate_limit_filter.rst]: broken link will update.

File: test/common/http/conn_manager_utility_test.cc [hunk 3]
```diff
@@ -341,4 +341,22 @@ TEST_F(ConnectionManagerUtilityTest, MutateResponseHeadersReturnXRequestId) {
...
+      {"x-forwarded-for", fmt::format("{0},{1}", first_address, second_address)}};
+  EXPECT_EQ(second_address, ConnectionManagerUtility::getLastAddressFromXFF(request_headers));
+}
+
+TEST_F(ConnectionManagerUtilityTest, EmptyXFF) {
+  HeaderMapImpl request_headers;
+  EXPECT_EQ("", ConnectionManagerUtility::getLastAddressFromXFF(request_headers));
+}
+
+TEST_F(ConnectionManagerUtilityTest, OneAddressInXff) {
```
Comment (ccaraman) [inline test/common/http/conn_manager_utility_test.cc]: Fix capitalization to be consistent.

File: docs/configuration/http_filters/rate_limit_filter.rst (hunk 2 above)
Comment (ccaraman) [inline docs/configuration/http_filters/rate_limit_filter.rst]: Updated.

File: docs/configuration/http_filters/rate_limit_filter.rst (hunk 1 above)
Comment (ccaraman) [inline docs/configuration/http_filters/rate_limit_filter.rst]: Updated documents and ratelimit.cc

File: include/envoy/http/filter.h [hunk 4]
```diff
@@ -101,6 +101,11 @@ class StreamFilterCallbacks {
    * put into the access log.
    */
   virtual AccessLog::RequestInfo& requestInfo() PURE;
+
+  /**
+   * @return the trusted address for the connection.
+   */
+  virtual const std::string& address() PURE;
```
Comment (mattklein123) [inline include/envoy/http/filter.h]: Perhaps downstreamAddress() ?

File: source/common/http/conn_manager_impl.h [hunk 5]
```diff
@@ -328,6 +329,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     void encodeTrailers(ActiveStreamEncoderFilter* filter, HeaderMap& trailers);
     void maybeEndEncode(bool end_stream);
     uint64_t streamId() { return stream_id_; }
+    const std::string& address() { return address_; }
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: I don't think this is needed

File: source/common/http/conn_manager_impl.h [hunk 6]
```diff
@@ -366,6 +368,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     std::list<std::function<void()>> reset_callbacks_;
     State state_;
     AccessLog::RequestInfoImpl request_info_;
+    std::string address_;
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: downstream_address_ or whatever you choose

File: source/common/http/filter/ratelimit.cc [hunk 7]
```diff
@@ -28,7 +29,8 @@ void ServiceToServiceAction::populateDescriptors(const Router::RouteEntry& route
 
 void RequestHeadersAction::populateDescriptors(const Router::RouteEntry& route,
                                                std::vector<::RateLimit::Descriptor>& descriptors,
-                                               FilterConfig&, const HeaderMap& headers) {
+                                               FilterConfig&, const HeaderMap& headers,
+                                               StreamDecoderFilterCallbacks&) {
   std::string header_value = headers.get(header_name_);
```
Comment (mattklein123) [inline source/common/http/filter/ratelimit.cc]: const std::string& while you are here

File: test/common/http/conn_manager_utility_test.cc (hunk 3 above)
Comment (ccaraman) [inline test/common/http/conn_manager_utility_test.cc]: Fixed

File: source/common/http/conn_manager_impl.h (hunk 5 above)
Comment (ccaraman) [inline source/common/http/conn_manager_impl.h]: Yes, it will be removed.

File: source/common/http/filter/ratelimit.cc (hunk 7 above)
Comment (ccaraman) [inline source/common/http/filter/ratelimit.cc]: Fixed.

File: source/common/http/conn_manager_impl.h (hunk 6 above)
Comment (ccaraman) [inline source/common/http/conn_manager_impl.h]: Changed to downstream_address_.

File: test/common/http/filter/ratelimit_test.cc [hunk 8]
```diff
@@ -1,3 +1,4 @@
+#include "common/common/empty_string.h"
```
Comment (mattklein123) [inline test/common/http/filter/ratelimit_test.cc]: nit: alpha order

Comment (ccaraman) [inline test/common/http/filter/ratelimit_test.cc]: Fixed.

File: source/common/http/conn_manager_utility.cc [hunk 9]
```diff
@@ -128,4 +129,15 @@ void ConnectionManagerUtility::mutateResponseHeaders(Http::HeaderMap& response_h
   }
 }
 
+std::string
```
Comment (RomanDzhabarov) [inline source/common/http/conn_manager_utility.cc]: i'd add this to http::utility where we do manipulations with headers

Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: Yeah that makes sense, let's move code/tests there.

Comment (ccaraman) [inline source/common/http/conn_manager_utility.cc]: That does make more sense. I'll move it there.

**Pull request merged**